comp/dogstatsd/server/batch.go

```diff
@@ -103,15 +103,15 @@ func (s *shardKeyGeneratorPerOrigin) Generate(sample metrics.MetricSample, shard
 	// We fall back on the generic sharding if:
 	// - the sample has a custom cardinality
 	// - we don't have the origin
-	if sample.OriginInfo.Cardinality != "" || (sample.OriginInfo.ContainerIDFromSocket == "" && sample.OriginInfo.LocalData.PodUID == "" && sample.OriginInfo.LocalData.ContainerID == "") {
+	if sample.OriginInfo.Cardinality != "" || (sample.OriginInfo.LocalData.ProcessID == 0 && sample.OriginInfo.LocalData.PodUID == "" && sample.OriginInfo.LocalData.ContainerID == "") {
 		return s.shardKeyGeneratorBase.Generate(sample, shards)
 	}
 
 	// Otherwise, we isolate the samples based on the origin.
 	i, j := uint64(0), uint64(0)
 	i, j = murmur3.SeedStringSum128(i, j, sample.OriginInfo.LocalData.PodUID)
 	i, j = murmur3.SeedStringSum128(i, j, sample.OriginInfo.LocalData.ContainerID)
-	i, _ = murmur3.SeedStringSum128(i, j, sample.OriginInfo.ContainerIDFromSocket)
+	i, _ = murmur3.SeedStringSum128(i, j, strconv.Itoa(int(sample.OriginInfo.LocalData.ProcessID)))
```

Comment:
This code historically avoids making allocations, and it looks like we don't have an automated SMP test to measure memory cost of handling dogstatsd traffic, so perhaps a manual benchmark is needed. Or maybe just xoring the process id into the hash would be good enough.

Reply:
Thanks! I've done the modification here: https://github.com/DataDog/datadog-agent/pull/33414/commits/25bdc5093d687c01693b9be0a67cc636c4cf4e45 👍

Reply:
FTR from Slack: metrics from a single container need to be aggregated together, and thus must be sent to the same shard. Multiple processes can belong to the same container, and using PID as a part of the sharding key would send them to different samplers, breaking aggregation. We cannot use anything more fine-grained than a container id here, unfortunately.

Reply:
Seen together, this feature has been removed in https://github.com/DataDog/datadog-agent/pull/34299 and this PR has been rebased on it.